torch/csrc/jit/type.cpp

```diff
@@ -85,4 +85,48 @@ ListTypePtr ListType::ofFloats() {
   return value;
 }
 
+at::optional<TypePtr> unifyTypes(const TypePtr& t1, const TypePtr& t2) {
+  //cases that t1 == t2, or t1 is a type refinement of t2 and vice versa
+  if (t1->isSubtypeOf(t2)) {
```

Comment:
Right now Tuple is considered a subtype of List, but they are not represented in the interpreter in the same way (this is perhaps a bad idea, but not something easily changed right now). The problem with this code is that it will allow a tuple to turn into a list without inserting a list construct node. The interpreter would then fail. I think for now it would be much safer to just check type equality after calling `unshapedType`. I believe that this is basically equivalent to this code but would not allow tuple->list which would cause the previously mentioned problem.